src/Juvix/Core/HR/Parser.hs

```diff
@@ -0,0 +1,134 @@
+module Juvix.Core.HR.Parser where
+
+import Data.Functor.Identity
+import Juvix.Core.HR.Types
+import Juvix.Core.Types
+import Juvix.Core.Usage
+import Juvix.Library hiding ((<|>), try)
+import Text.Parsec hiding (try)
+import Text.ParserCombinators.Parsec
+import Text.ParserCombinators.Parsec.Expr
+import Text.ParserCombinators.Parsec.Language hiding (reservedNames, reservedOpNames)
+import qualified Text.ParserCombinators.Parsec.Token as Token
+import Prelude (String)
+
+baseReservedNames ∷ [String]
+baseReservedNames =
+  [ "*", -- sort
+    "[Π]", -- function type
+    "w" -- omega
+  ]
+
+baseReservedOpNames ∷ [String]
+baseReservedOpNames =
+  [ "\\", -- lambda
+    "@", -- TODO: remove me, necessary for annotation parsing at the moment
+    ":", -- type & usage annotation
+    "->" -- arrow
+  ]
+
+generateParser ∷ ∀ primTy primVal. Parameterisation primTy primVal → (String → Maybe (Term primTy primVal))
+generateParser parameterisation =
```

Comment:
Rather hard to read due to how the code formatter handles this. 

However in the future, we should fix our code formatter to add newlines if it sees ∷ in a let block.